firedrake/extrusion_extend.py

```diff
@@ -0,0 +1,42 @@
+from firedrake import FunctionSpace, Function, DirichletBC
+
+__all__ = ['ExtrudedExtendFunction']
+
+def ExtrudedExtendFunction(mesh, fcn, extend_type=None):
+    """On an extruded mesh, extend a function by a constant in the extruded
+    direction.  The returned function uses the degree 0 'R' space in the
+    extruded direction.
+
+    :arg mesh:           the extruded mesh
+
+    :arg fcn:            the function to extend
+
+    ``extend_type`` has three cases:
+
+    ``None``
+        ``fcn`` is defined on the base mesh and the returned Function has
+        the same values in the extruded direction.
+
+    ``"top"`` or ``"bottom"``
+        ``fcn`` is defined on the extruded mesh.  The values of the returned
+        Function come from the values of ``fcn`` on the ``"top"``/``"bottom"``
+        boundary of the extruded mesh.
+    """
+
+    # FIXME need to use fcn.function_space() to get element; e.g.
+    #     Q3D = FunctionSpace(mesh, fcn.function_space().ufl_element, vfamily='R', vdegree=0)
+    # does not work; note the base mesh is mesh._base_mesh
+
+    Q3D = FunctionSpace(mesh, 'CG', 1, vfamily='R', vdegree=0)  # FIXME
+    f3D = Function(Q3D)
```

Comment:
Would be good to have an optional argument for a preallocated f3D, for time stepping.

Reply:
?

Reply:
Rather than making a new Function, allow it to be passed in. 

So 
```
def ExtudedExtend(..., target=None):
  If target is not None:
       assert target.ufl_element() == Q3D.ufl_element()
        f3d = target
    else:
        f3d = Function(Q3D)
```

Reply:
This makes sense, although I'm not completely sure what a test case would be.

Reply:
I understand this case and it is in the passed test cases.